### PR TITLE
SONARPY-2336 Ensure ClassDescriptor and FunctionDescriptor fullyQualifiedName are not nullable

### PR DESCRIPTION
**Pull request by guillaume-dequenne-sonarsource**
[SONARPY-2339](https://sonarsource.atlassian.net/browse/SONARPY-2339)

Following this PR, there are possible null arguments to `ClassDescriptor#fullQualifiedName` and `FunctionDescriptor#fullQualifiedName` in `DescriptorUtils`. These methods are no longer used in production code, and are only used in tests (mostly in `ProjectLevelSymbolTableTest` through `ProjectLevelSymbolTable#from`). [SONARPY-2346](https://sonarsource.atlassian.net/browse/SONARPY-2346) has been created to handle this.

[SONARPY-2339]: https://sonarsource.atlassian.net/browse/SONARPY-2339?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SONARPY-2346]: https://sonarsource.atlassian.net/browse/SONARPY-2346?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: python-frontend/src/main/java/org/sonar/python/types/v2/ClassType.java [hunk 1]
```diff
@@ -210,6 +209,10 @@ public String name() {
     return name;
   }
 
+  public String fullyQualifiedName() {
+    return fullyQualifiedName;
+  }
```
Comment (thomas-serre-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/v2/ClassType.java:214]: _nit_ What about adding `@Nonnull` here?
```suggestion
  @Nonnull 
  public String fullyQualifiedName() {
    return fullyQualifiedName;
  }
```

File: python-frontend/src/main/java/org/sonar/python/semantic/v2/converter/PythonTypeToDescriptorConverter.java
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/v2/converter/PythonTypeToDescriptorConverter.java:1]: These are actually not the same thing, and there is a side effect to doing this when dealing with aliases:
```
# lib.py
class MyClass():
        def foo(self): ...
my_alias = MyClass
```
In the example, `my_alias` represents a class type for which we create a descriptor with name `my_alias` and FQN `lib.my_alias`. If I make the change, we now have 2 distinct type for the FQN `lib.MyClass`,  and that crashes `AmbiguousDescriptor`. 

One easy way to fix it is not to re-export class aliases (by filtering out class types whose symbol has a different name than the class type for example), but that leads to FNs.
A cleaner fix would be to correctly represent that `my_alias` is an `AliasDescriptor` pointing to `MyClass`. I think it's a bit too big for the current scope though, so I created SONARPY-2352 to fix this.

In the meantime, I suggest we keep the current behavior, which I documented through a new test.

**Comment by guillaume-dequenne-sonarsource** [comment]
> LGTM, great work 💪 !
> 
> I left many minor comments, and I think it is worth it to have a second round of review.
> 
> On the JIRA side, it looks a bit unclear to me. In my opinion you are tackling 3 tickets in this PR :
> 
> * [SONARPY-2336](https://sonarsource.atlassian.net/browse/SONARPY-2336) as written in the PR title
> * [SONARPY-2339](https://sonarsource.atlassian.net/browse/SONARPY-2339) as commented in the PR
> * SONARPY-???? which does not exist and is meant to convert ***TypeTest to V2 (see comments in the review for a clearer explanation).
> 
> I'm okay with tackling several tickets in the same PR, but I'm unsure if the jira automation will work in this case, and I would prefer to see the 3 jira tickets in the title of the PR. It could be worth it to settle this so the development process covers this case.

I think the 3rd ticket you're referring to matches SONARPY-2346. I'm tackling a small part of it here because it's needed to repair the otherwise broken tests on `ClassDescriptor` otherwise. I probably should have squashed the 3rd commit into the second one (I believe I simply forgot to do it...).

When dealing with multiple tickets in a single PR, I personally don't like having them all in the PR title (I tend to think it feels bloated then). I tend to have the "main" ticket as a PR title, and the prerequisite as a separate commit. Indeed I don't think Jira automation works in this case. We can probably revisit this way of doing so that it works better with automation though.

File: python-frontend/src/main/java/org/sonar/python/types/v2/ClassType.java (hunk 1 above)
Comment (thomas-serre-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/v2/ClassType.java:214]: makes sense, thanks!

File: python-frontend/src/main/java/org/sonar/python/semantic/v2/converter/PythonTypeToDescriptorConverter.java
[the remark this answers is not in the document]
Comment (thomas-serre-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/v2/converter/PythonTypeToDescriptorConverter.java:1]: thanks for the explanation! yep, let's keep it like this :)

**Review by thomas-serre-sonarsource — APPROVED** [review]
LGTM! 

Thanks for the explanations along the fixes :muscle: 

I ll add a point to the weekly meeting to confirm and ensure everyone does the same thing when one PR is created for several tickets.

**Pull request merged**